### PR TITLE
Add doOnEach variant similar to doFirst&doOnCancel + Context

### PR DESCRIPTION
**Pull request by simonbasle**
This commit adds doOnEachSubscriptionLifecycle, a dual to doOnEach
that allows listening to SUBSCRIBE, ON_SUBSCRIBE and CANCEL events in a
single lambda.

Additionally, contrary to doFirst/doOnCancel/doOnSubscribe it exposes
the Context.

Extending Signal with these new SignalType could have induced breaking
of existing code that doesn't strictly check the Signal's type when
processing it and using its getters. So this new operator doesn't expose
a Signal but works with a `BiConsumer<SignalType, Context>`. This is
assumed sufficient, as the "data" associated to such signals is of
limited use (basically Subscription for ON_SUBSCRIBE).

### CODE REVIEW

File: reactor-core/src/main/java/reactor/core/publisher/FluxDoOnEach.java [hunk 1]
```diff
@@ -100,13 +115,19 @@ public void request(long n) {
 		@Override
 		public void cancel() {
 			s.cancel();
+			if (onSubscriptionSignal != null) {
+				onSubscriptionSignal.accept(SignalType.CANCEL, cachedContext);
```
Comment (ericbottard) [inline reactor-core/src/main/java/reactor/core/publisher/FluxDoOnEach.java:119]: Noob question: what happens if the callback throws an exception, and should we care about it?

File: reactor-core/src/main/java/reactor/core/publisher/ImmutableSignal.java [hunk 2]
```diff
@@ -82,49 +82,38 @@ public Context getContext() {
 	}
 
 	@Override
-	public boolean equals(@Nullable Object o) {
+	public boolean equals(Object o) {
```
Comment (ericbottard) [inline reactor-core/src/main/java/reactor/core/publisher/ImmutableSignal.java:85]: Can you elaborate on these apparent semantics change?

Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/publisher/ImmutableSignal.java]: my first iteration was adding the SUBSCRIBE and CANCEL cases to the `Signal` class, but that is likely to cause issues with existing open-ended `if-elseif-else` styles of dealing with the `Signal`. I've since reverted.

File: reactor-core/src/main/java/reactor/core/publisher/FluxDoOnEach.java (hunk 1 above)
Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/publisher/FluxDoOnEach.java:119]: not a noob question at all, it should indeed be protected.

**Comment by simonbasle** [comment]
@reactor/core-team I'd like the opinion of the whole team on the API (naming, limitations, etc...) 🙇

File: reactor-core/src/main/java/reactor/core/publisher/OptimizableOperator.java [hunk 3]
```diff
@@ -35,7 +35,13 @@
 
 	/**
 	 * Allow delegation of the subscription by returning a {@link CoreSubscriber}, or force
-	 * subscription encapsulation by returning null. This can be used in conjunction with {@link #nextOptimizableSource()}
+	 * subscription encapsulation by returning null.
+	 * <br/>
+	 * Caller MUST catch exceptions and report them to the most relevant {@link org.reactivestreams.Subscriber},
+	 * generally using {@link Operators#reportThrowInSubscribe(CoreSubscriber, Throwable)},
```
Comment (ericbottard) [inline reactor-core/src/main/java/reactor/core/publisher/OptimizableOperator.java:41]: Looking at the current code base, some implementors of this API actually do something like
```java
Operators.error(actual, Operators.onOperatorError(e, actual.currentContext()));
```

Is that OK? (If not, to be dealt with in another PR, but wanted to raise the point)

**Comment by bsideup** [comment]
@simonbasle TBH I am on the fence about adding Context to various operators. It only creates more work, while IMO it would be better to make it easy to _access_ the context (e.g. `transformDeferred(ctx -> {})` that we discussed)

File: reactor-core/src/main/java/reactor/core/publisher/OptimizableOperator.java (hunk 3 above)
Comment (bsideup) [reply reactor-core/src/main/java/reactor/core/publisher/OptimizableOperator.java:41]: @ericbottard could you please point me at any place where `Operators.error(actual, Operators.onOperatorError` in `subscribe` is used instead of `reportThrowInSubscribe`?

Comment (ericbottard) [reply reactor-core/src/main/java/reactor/core/publisher/OptimizableOperator.java:41]: From "Show call hierarchy" in IJ:
```
MonoIgnorePublisher
FluxReplay
MonoDelayUntil
MonoFromPublisher
```

Comment (bsideup) [reply reactor-core/src/main/java/reactor/core/publisher/OptimizableOperator.java:41]: That's because `source` is `Publisher`, and `subscribeOrReturn` is defined next to `subscribe()` + final and will never throw

Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/publisher/OptimizableOperator.java:41]: yeah, I saw that hence the `generally`. it is ok for now but would make sense in aligning in a separate issue I think.

**Comment by simonbasle** [comment]
> @simonbasle TBH I am on the fence about adding Context to various operators. It only creates more work, while IMO it would be better to make it easy to _access_ the context (e.g. `transformDeferred(ctx -> {})` that we discussed)

I see this change as complementary. It is a way to expose another _range_ of events in a single operator. Another avenue to explore would be to expose `FluxPeek`/`SignalPeek` in a public API (e.g. with a default base implementation like the `BaseSubscriber`)

File: reactor-core/src/main/java/reactor/core/publisher/OptimizableOperator.java (hunk 3 above)
Comment (bsideup) [reply reactor-core/src/main/java/reactor/core/publisher/OptimizableOperator.java:41]: There is nothing to align, we intentionally did not change these 4 :)

Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/publisher/OptimizableOperator.java:41]: ok then 👍

**Comment by simonbasle** [comment]
@bsideup still on the fence about this one ? especially with the `scope` also coming around the corner?

**Comment by simonbasle** [comment]
after discussing this internally, we'll focus on `deferWithContext`/`transformDeferred`-with-context
